src/DynamoCore/Core/PathManager.cs

```diff
@@ -0,0 +1,280 @@
+            }
+        }
+
+        /// <summary>
+        /// Given an initial file path with the file name, resolve the full path
+        /// to the target file. The search happens in the following order:
+        /// 
+        /// 1. If the provided file path is valid and points to an existing file, 
+        ///    the file path is return as-is.
+        /// 2. The file is searched alongside DynamoCore.dll for a match.
+        /// 3. The file is searched within AdditionalResolutionPaths.
+        /// 4. The search is left to system path resolution.
+        /// 
+        /// </summary>
+        /// <param name="library">The initial library file path.</param>
+        /// <returns>Returns true if the requested file can be located, or false
+        /// otherwise.</returns>
+        /// 
+        public bool ResolveLibraryPath(ref string library)
+        {
+            if (File.Exists(library)) // Absolute path, we're done here.
+                return true;
+
+            library = LibrarySearchPaths(library).FirstOrDefault(File.Exists);
+            return library != default(string);
+        }
+
+        #endregion
+
+        #region Public Class Operational Methods
+
+        /// <summary>
+        /// Constructs an instance of PathManager object.
+        /// </summary>
+        /// <param name="pathResolver">Reference of an IPathResolver object that
+        /// supplies additional path information. This argument is optional.</param>
+        /// 
+        internal PathManager(IPathResolver pathResolver)
+        {
+            // This method is invoked in DynamoCore.dll, dynamoCorePath 
+            // represents the directory that contains DynamoCore.dll.
+            var dynamoCorePath = Assembly.GetExecutingAssembly().Location;
+            dynamoCoreDir = Path.GetDirectoryName(dynamoCorePath);
+
+            if (!File.Exists(Path.Combine(dynamoCoreDir, "DynamoCore.dll")))
+            {
+                throw new Exception("Dynamo's core path could not be found. " +
+                    "If you are running Dynamo from a test, try specifying the " +
+                    "Dynamo core location in the DynamoBasePath variable in " +
+                    "TestServices.dll.config.");
+            }
+
+            // Current user specific directories.
+            userDataDir = CreateFolder(GetUserDataFolder());
+
+            userDefinitions = CreateFolder(Path.Combine(userDataDir, "definitions"));
+            logDirectory = CreateFolder(Path.Combine(userDataDir, "Logs"));
+            packagesDirectory = CreateFolder(Path.Combine(userDataDir, "packages"));
+            preferenceFilePath = Path.Combine(userDataDir, "DynamoSettings.xml");
+
+            // Common directories.
+            commonDataDir = CreateFolder(GetCommonDataFolder());
+
+            commonDefinitions = CreateFolder(Path.Combine(commonDataDir, "definitions"));
+            samplesDirectory = GetSamplesFolder(commonDataDir);
+
```

Comment:
That's a good point, @pboyer! Technically the `PathManager` does not have to create these folders internally, they should come pre-configured with the installer. As for unit tests, you're right that folder creation should not become a side effect and that the clients (unit tests) should be allowed to set this to other locations. For that, I have made a new internal task to be done this sprint:

[MAGN-6605](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6605) IPathResolver should allow clients to override data folder locations

This will expose fully configurable data folder locations through `IPathResolver`.